cpp/src/arrow/compute/exec/options.h

```diff
@@ -106,21 +106,32 @@ class ARROW_EXPORT ProjectNodeOptions : public ExecNodeOptions {
   std::vector<std::string> names;
 };
 
-/// \brief Make a node which aggregates input batches, optionally grouped by keys.
+/// \brief Make a node which aggregates input batches, optionally grouped by keys and
+/// optionally segmented by segment-keys. Both keys and segment-keys determine the group.
+/// However segment-keys are also used for determining grouping segments, which should be
+/// large, and allow streaming a partial aggregation result after processing each segment.
```

Comment:
Ideally all of your keys would be segment keys, the smaller the segments the better I would think.  Do you agree or do you imagine it otherwise?  Also, this comment implies but does not explicitly mention that the incoming data must be ordered by the segment keys.

Reply:
The definition is: when moving from one row to the next, if the tuple of segment keys is constant then both rows are in the same segment, whereas if the tuple changes in any way (not necessarily in lexicographic order) then they are in different segments - see also [this post](https://github.com/apache/arrow/pull/14352#issuecomment-1272400060).

There are use cases for any segment size. If the segments are smaller, there will be more group-by processing overhead, at least due to processing of more batches and due to reinitializing the states on each segment crossing, and the resulting stream will generate more batches and with lower latency.

Reply:
> and the resulting stream will generate more batches and with lower latency.

We can maybe collect outgoing data to ensure we aren't sending anything out of this node that is too tiny.  That can be left for a future PR.

I see your point on state initialization though that is generally a pretty cheap operation (I think).  Mainly I don't want to have to burden users with trying to optimize segment sizes for performance.  That feels like a problem we can solve internally and they should just setup their segments to match their data.

Reply:
> The definition is: when moving from one row to the next, if the tuple of segment keys is constant then both rows are in the same segment, whereas if the tuple changes in any way (not necessarily in lexicographic order) then they are in different segments - see also https://github.com/apache/arrow/pull/14352#issuecomment-1272400060.

Also, thanks for this explanation.  I agree it sounds like a useful generalization.  Thanks!

Reply:
> We can maybe collect outgoing data to ensure we aren't sending anything out of this node that is too tiny. That can be left for a future PR.

Agreed.

> I see your point on state initialization though that is generally a pretty cheap operation (I think). Mainly I don't want to have to burden users with trying to optimize segment sizes for performance. That feels like a problem we can solve internally and they should just setup their segments to match their data.

I agree we could deal with size optimizations internally - I'd prefer in a separate jira. It's actually the output batch size being optimized because the segments, and their sizes, are determined by the input data. Note that the cost of a small segment sizes may not be trivial as it also includes the overhead of finishing (see [this post](https://github.com/apache/arrow/pull/14352#discussion_r1019650164)), at least with the current design. It also includes the usual processing cost due to outputting more batches.

Reply:
https://issues.apache.org/jira/browse/ARROW-18312